### PR TITLE
tree: add and use create utils

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

Add some simple tree inspired tree creation utilities.

This focuses on two formats:

1. "insertable" which can be json compatible if you special case handles and works well for unambiguous schema. Similar to using schema constructors, but can construct cursors, as well as deal with unions and undefined at the root.
2. "VerboseTree": handles ambiguous schema, and is a simple self describing format which is good for persisting, even without schema. Think of it as JsonableTree for simple-tree.

Not included are APIs to clone trees into these formats for export: such APIs are planned for the future, and will allow round tripping data.

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

### CODE REVIEW

File: packages/dds/tree/src/simple-tree/api/verboseTree.ts [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+ * This format allows for all simple-tree compatible trees to be represented.
+ *
+ * Unlike `JsonableTree`, leaf nodes are not boxed into node objects, and instead have their schema inferred from the value.
+ * Additionally, sequence fields can only occur on a node that has a single sequence field (with the empty key) replicating the behavior of simple-tree ArrayNodes.
+ */
+export interface VerboseTreeNode<THandle = IFluidHandle> {
+	/**
+	 * The meaning of this node.
+	 * Provides contexts/semantics for this node and its content.
+	 * Typically used to associate a node with metadata (including a schema) and source code (types, behaviors, etc).
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/api/verboseTree.ts:86]: Nit: this last line probably belongs in a `@remarks`, since it describes potential usage. An example would also probably be useful.

File: packages/dds/tree/src/simple-tree/api/verboseTree.ts [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+}
+
+/**
+ * Options for how to interpret a `VerboseTree<TCustom>` when schema information is available.
+ */
+export interface ParseOptions<TCustom> {
+	/**
+	 * Fixup custom input formats.
+	 * @remarks
+	 * Main usage is translate some JSON compatible handle format into actual IFluidHandles.
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/api/verboseTree.ts:111]: Nit ```suggestion
	 * Main usage is translate some JSON-compatible handle format into actual IFluidHandles.
```

Comment (Josmithr) [reply packages/dds/tree/src/simple-tree/api/verboseTree.ts:111]: An example of this scenario would probably be good too. Also, currently the toJsonSchema logic throws if it encounters a Fluid Handle. If we have a reasonable representation we can support round-trip, we can probably lift that restriction.

File: packages/dds/tree/src/simple-tree/api/verboseTree.ts [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+	 */
+	valueConverter(data: VerboseTree<TCustom>): TreeLeafValue | VerboseTreeNode<TCustom>;
+	/**
+	 * If true, interpret the input keys of object nodes as stored keys.
+	 * If false, interpret them as property keys.
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/api/verboseTree.ts:116]: I think in other places we refer to these as "view" keys. Not sure which terminology is used more, but it would be good for us to settle on a single term.

File: packages/dds/tree/src/simple-tree/objectNode.ts [hunk 4]
```diff
@@ -321,9 +321,18 @@ export function objectSchema<
 
 	class CustomObjectNode extends CustomObjectNodeBase<T> {
 		public static readonly fields: ReadonlyMap<string, FieldSchema> = new Map(
-			[...flexKeyMap].map(([key, value]) => [key as string, value.schema]),
+			Array.from(flexKeyMap, ([key, value]) => [key as string, value.schema]),
 		);
 		public static readonly flexKeyMap: SimpleKeyMap = flexKeyMap;
+		public static readonly storedKeyToPropertyKey: ReadonlyMap<FieldKey, string> = new Map<
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/objectNode.ts:327]: I believe we already cache such a mapping somewhere. We should dedupe. @alexvy86 @noencke do either of you remember where that existing mapping lives?

File: packages/dds/tree/src/test/scalableTestTrees.ts [hunk 5]
```diff
@@ -51,9 +52,19 @@ const wideBuilder = new SchemaBuilderBase(FieldKinds.required, {
 });
 
 export const wideRootSchema = wideBuilder.object("WideRoot", {
-	foo: FlexFieldSchema.create(FieldKinds.sequence, [leaf.number]),
+	[EmptyKey]: FlexFieldSchema.create(FieldKinds.sequence, [leaf.number]),
 });
 
+const sf = new SchemaFactory("scalable");
+export class LinkedList extends sf.objectRecursive("linkedList", {
```
Comment (Josmithr) [inline packages/dds/tree/src/test/scalableTestTrees.ts:59]: Docs for these would be good, since they are exported and intended for re-use.

File: packages/dds/tree/src/test/scalableTestTrees.ts [hunk 6]
```diff
@@ -95,12 +101,12 @@ export function makeDeepContent(
  */
 export function makeWideContentWithEndValue(
 	numberOfNodes: number,
-	endLeafValue: number,
+	endLeafValue?: number,
```
Comment (Josmithr) [inline packages/dds/tree/src/test/scalableTestTrees.ts:104]: We should update the docs to explain what it means for this to not be provided.

File: packages/dds/tree/src/test/scalableTestTrees.ts [hunk 7]
```diff
@@ -231,7 +238,7 @@ export function readWideFlexTree(tree: FlexTreeView<typeof wideSchema.rootFieldS
 	let sum = 0;
 	let nodesCount = 0;
 	const root = tree.flexTree;
-	const field = root.content.foo;
+	const field = root.content[""];
```
Comment (Josmithr) [inline packages/dds/tree/src/test/scalableTestTrees.ts:241]: Should we use `EmptyKey` here?

File: packages/dds/tree/src/simple-tree/api/verboseTree.ts (hunk 2 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/simple-tree/api/verboseTree.ts:111]: There is currently no way to encode a fluid handle using our public APIs. If there was using it would tend to result in GC issues as the data would disappear despite your custom reference to it existing.

Realistically there are 4 things you can currently do:
1. throw
2. leave handles as is (and not be json compatible)
3. Provide some domain specific encoding of handles (ex: I know this handle points to Foo, so I'll encode it as "container/foo")
4. use internal APIs to encode the handle somehow. One option is to make a non-round triable but at least comparable encoding (ex: pull out the handle absolute path or hash), the other would be to use the alpha+legacy fluid serializer and its round trippable format (the internal format, ISerializedHandle )

Regardless, you are right these docs need to be better.

File: packages/dds/tree/src/simple-tree/api/verboseTree.ts (hunk 3 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/simple-tree/api/verboseTree.ts:116]: Do you have an example of such a place? FieldProps,key's docs calls them "property keys". I think that aligns with our API which uses them as the keys for properties on object nodes (ex: Object.prototype.hasOwnProperty calles them properties, and TypeScript types the parameter "key" parameter  as "PropertyKey")

Comment (CraigMacomber) [reply packages/dds/tree/src/simple-tree/api/verboseTree.ts:116]: Note that the term "view schema" is generally not used in our public APi as its the only public schema system we have. Nearly everything thats public is implicitly "view".

File: packages/dds/tree/src/simple-tree/objectNode.ts (hunk 4 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/simple-tree/objectNode.ts:327]: We have the inverse of this (lthe line above it) but nothing in this direction that I'm aware of.

There is another open PR that also adds it ( https://github.com/microsoft/FluidFramework/pull/22043/files#diff-978f95b1fbda78856fe9c4c79e72a61dce91cf7bbdd8b3fcf0e6c9dbf6685165R55 ), but I don't think it exists in main.

File: packages/dds/tree/src/test/forestTestSuite.ts [hunk 8]
```diff
@@ -1133,6 +1133,7 @@ export function testForest(config: ForestTestConfiguration): void {
 			});
 			it("when moving the last node in the field", () => {
 				const schemaFactory = new SchemaFactory("moving");
+				// TODO: this test does not seem to be in schema. Its moving from a required field to another required field.
```
[the remark this answers is not in the document]
Comment (CraigMacomber) [reply packages/dds/tree/src/test/forestTestSuite.ts:1136]: Its in a test, only impacts this specific code so anyone impacted by it can see the TODO making having external tracking for it very low value. Prabably more than half our 480+ tree TODOs are higher priority. I don't think adding it to our bug backlog would be helpful (its not a user impacting issue) and we are not tracking tech debt items this small individually.

Also this code is guaranteed to get revisited soon as part of removing typedJsonCursor (which I tried to do here, but failed since the schema and code don't align)

File: packages/dds/tree/src/simple-tree/objectNode.ts (hunk 4 above)
Comment (Josmithr) [reply packages/dds/tree/src/simple-tree/objectNode.ts:327]: Ah, that PR is what I was thinking of. Okay. I guess we should just make sure we don't accidentally duplicate solutions between the PRs.

File: packages/dds/tree/src/simple-tree/api/verboseTree.ts (hunk 2 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/simple-tree/api/verboseTree.ts]: docs improved. Will likley need more before this goes public, but I think this is enough for our own internal test use in the short term.

File: packages/dds/tree/src/simple-tree/api/verboseTree.ts (hunk 3 above)
Comment (Josmithr) [reply packages/dds/tree/src/simple-tree/api/verboseTree.ts:122]: Yeah, a brief search didn't reveal the "view key" terminology in public APIs. I was thinking it ended up getting surfaced through `FieldProps` / `FieldSchema` / `treeNodeApi`, but that doesn't seem to be the case.

Still, we may want to do an audit of our non-public docs to unify that terminology. Do we have a glossary of terms anywhere?

File: packages/dds/tree/src/simple-tree/api/create.ts [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+import {
+	applySchemaToParserOptions,
+	cursorFromVerbose,
+	type ParseOptions,
+	type VerboseTree,
+	type VerboseTreeNode,
+} from "./verboseTree.js";
+
+/**
+ * Construct tree content compatible with a field defined by the provided `schema`.
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/api/create.ts:41]: Nit (was a bit hard to parse):
```suggestion
 * Construct tree content that is compatible with the field defined by the provided `schema`.
```

File: packages/dds/tree/src/simple-tree/api/create.ts [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+}
+
+/**
+ * Construct tree content compatible with a field defined by the provided `schema`.
+ * @param schema - The schema for what to construct. As this is an {@link ImplicitFieldSchema}, a {@link FieldSchema}, {@link TreeNodeSchema} or {@link AllowedTypes} array can be provided.
+ * @param data - The data used to construct the field content.
+ * @remarks
+ * When providing a {@link TreeNodeSchemaClass},
+ * this is the same as invoking its constructor except that an unhydrated node can also be provided and the returned value is a cursor.
+ * When `undefined` is provided (for an optional field), `undefined` is returned.
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/api/create.ts:71]: Does this apply to `identifier` fields as well? Should this say "for fields with defaults"?

Comment (Josmithr) [reply packages/dds/tree/src/simple-tree/api/create.ts:71]: (Same with the docs for the previous method, if so)

File: packages/dds/tree/src/simple-tree/api/create.ts [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+		},
+		...options,
+	};
+	const schemalessConfig = applySchemaToParserOptions(schema, config);
+	const cursor = cursorFromVerbose(data, schemalessConfig);
+	return createFromCursor(schema, cursor);
+}
+
+/**
+ * Creates an unhydrated simple-tree field from a cursor in nodes mode.
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/api/create.ts:144]: What does this do if the cursor is not in "nodes" mode? Should we document that?

File: packages/dds/tree/src/simple-tree/api/verboseTree.ts [hunk 12]
```diff
@@ -0,0 +1,415 @@
...
+	stringSchema,
+} from "../leafNodeSchema.js";
+import { toFlexSchema } from "../toFlexSchema.js";
+import { isObjectNodeSchema } from "../objectNodeTypes.js";
+import { walkFieldSchema } from "./tree.js";
+
+/**
+ * Verbose encoding of a {@link TreeNode} or {@link TreeValue}.
+ * @remarks
+ * This is verbose meaning that every {@link TreeNode} is a {@link VerboseTreeNode}.
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/api/verboseTree.ts:49]: Nit:
```suggestion
 * This is "verbose", meaning that every {@link TreeNode} is a {@link VerboseTreeNode}.
```

File: packages/dds/tree/src/simple-tree/api/verboseTree.ts [hunk 13]
```diff
@@ -0,0 +1,415 @@
...
+ * Any IFluidHandle values have been replaced by `THandle`.
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/api/verboseTree.ts:50]: Nit
```suggestion
 * Any {@link @fluidframework/core-interfaces#IFluidHandle} values have been replaced by `THandle`.
```

File: packages/dds/tree/src/simple-tree/api/verboseTree.ts [hunk 14]
```diff
@@ -0,0 +1,415 @@
...
+export type VerboseTree<THandle = IFluidHandle> =
+	| VerboseTreeNode<THandle>
+	| Exclude<TreeLeafValue, IFluidHandle>
+	| THandle;
+
+/**
+ * Verbose encoding of a {@link TreeNode}.
+ * @remarks
+ * This is verbose meaning that every {@link TreeNode} has an explicit `type` property, and `fields`.
+ * This allowed VerboseTreeNode to be unambiguous regarding which type each node is without relying on symbols or hidden state.
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/api/verboseTree.ts:63]: Nit
```suggestion
 * This allows VerboseTreeNode to be unambiguous regarding which type each node is without relying on symbols or hidden state.
```

File: packages/dds/tree/src/simple-tree/api/verboseTree.ts [hunk 15]
```diff
@@ -0,0 +1,415 @@
...
+ *
+ * Any IFluidHandle values have been replaced by `THandle`. If the `THandle` is JSON compatible, then this type is JSON compatible as well.
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/api/verboseTree.ts:65]: Nit
```suggestion
 * Any {@link @fluidframework/core-interfaces#IFluidHandle} values have been replaced by `THandle`. If the `THandle` is JSON-compatible, then this type is JSON-compatible as well.
```

File: packages/dds/tree/src/simple-tree/api/verboseTree.ts [hunk 16]
```diff
@@ -0,0 +1,415 @@
...
+ *
+ * @privateRemarks
+ * This type is only used for data which is copied into and out of the tree.
+ * When being copied out, its fine to have the data be mutable since its a copy.
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/api/verboseTree.ts:69]:
```suggestion
 * When being copied out, it's fine to have the data be mutable since its a copy.
```

File: packages/dds/tree/src/simple-tree/api/verboseTree.ts [hunk 17]
```diff
@@ -0,0 +1,415 @@
...
+}
+
+/**
+ * Options for how to interpret a `VerboseTree<TCustom>` when schema information is available.
+ */
+export interface ParseOptions<TCustom> {
+	/**
+	 * Fixup custom input formats.
+	 * @remarks
+	 * Main usage is to handle IFluidHandles.
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/api/verboseTree.ts:113]: Nit
```suggestion
	 * Main usage is to handle {@link @fluidframework/core-interfaces#IFluidHandle}s.
```

File: packages/dds/tree/src/simple-tree/api/verboseTree.ts [hunk 18]
```diff
@@ -0,0 +1,415 @@
...
+	 * When targeting JSON compatibility,
+	 * this may be by throwing an error or including a placeholder.
+	 * Since IFluidHandles are special references to FLuid data which is garbage collected when not referenced by the container for long enough,
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/api/verboseTree.ts:116]:
```suggestion
	 * Since IFluidHandles are special references to Fluid data which is garbage collected when not referenced by the container for long enough,
```

File: packages/dds/tree/src/simple-tree/core/treeNodeSchema.ts [hunk 19]
```diff
@@ -121,6 +121,16 @@ export interface TreeNodeSchemaCore<
 	out ImplicitlyConstructable extends boolean,
 	out Info = unknown,
 > {
+	/**
+	 * Unique (withing a document's schema) identifier used to associate nodes with their schema.
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/core/treeNodeSchema.ts:125]:
```suggestion
	 * Unique (within a document's schema) identifier used to associate nodes with their schema.
```

Comment (Josmithr) [reply packages/dds/tree/src/simple-tree/core/treeNodeSchema.ts:125]: Nit: Might also be a bit easier to read if you move the parens block after "identifier".

File: packages/dds/tree/src/simple-tree/core/treeNodeSchema.ts [hunk 20]
```diff
@@ -121,6 +121,16 @@ export interface TreeNodeSchemaCore<
...
+	 * @remarks
+	 * This is used when encoding nodes, and when decoding nodes to re-associate them with the schema.
+	 * Since this decoding may happen in a different version of the application (or even a different application all together),
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/core/treeNodeSchema.ts:128]: Nit
```suggestion
	 * Since this decoding may happen in a different version of the application (or even a different application altogether),
```

File: packages/dds/tree/src/test/scalableTestTrees.ts (hunk 6 above)
Comment (Josmithr) [reply packages/dds/tree/src/test/scalableTestTrees.ts:121]: The comment above this function could still use an update.

**Review by Josmithr — APPROVED** [review]
Left some documentation suggestions, but otherwise this looks good to me!

File: packages/dds/tree/src/test/feature-libraries/flex-map-tree/mapTreeNode.spec.ts [hunk 21]
```diff
@@ -102,7 +102,7 @@ describe("MapTreeNodes", () => {
 	it("can get their type", () => {
 		assert.equal(map.type, "Test.Map");
 		assert.equal(fieldNode.type, "Test.FieldNode");
-		assert.equal(object.type, "Test.Object");
+		assert.equal(object.mapTree.type, "Test.Object");
```
Comment (noencke) [inline packages/dds/tree/src/test/feature-libraries/flex-map-tree/mapTreeNode.spec.ts:105]:
```suggestion
		assert.equal(object.type, "Test.Object");
```

File: packages/dds/tree/src/simple-tree/api/create.ts (hunk 10 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/simple-tree/api/create.ts:71]: mapTreeFromNodeData has several issues in the identifier case. It has several TODOs and a couple bugs ( AB#9126 AB#9131 ) regarding the fact that we never validation anything even when opting in to validation which is the only reason identifier fields using defaults don't fail validation, and an additional TODO about not supporting defaults for the root case.

Even if it supported defaults for the root case, if you didn't provide the optional NodeKeyManager, it would just return undefined for an identifier and fail schema validation.

As those are all existing issue with production code this calls, and are tracked by several bugs and TODOs, I think not calling out those issues here might be ok (but we really need to get them fixed sometime soon!)

File: packages/dds/tree/src/simple-tree/api/create.ts (hunk 11 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/simple-tree/api/create.ts:144]: doing so is invalid, will cause undefined behavior which likely depends on the cursor implementation (for our less optimized cursors, you would get a nice error). Most cursor stuff is like that, so I think its fine. This design approach dates back to when our number one priority was minimal overhead on the data access paths so the production cursers have minimal validation and are specced to not require such validation to exist
